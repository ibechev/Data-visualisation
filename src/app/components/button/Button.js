import React from "react";
import PropTypes from "prop-types";

const Button = ({ children, noBody, bgc }) => {
  const buttonClass = `button ${bgc} ${noBody ? "no-body" : ""}`;

  return (
    <button type="button" className={buttonClass}>
      {children}
    </button>
  );
};

Button.defaultProps = {
  noBody: false,
  bgc: "primary"
};

Button.propTypes = {
  children: PropTypes.node.isRequired,
  noBody: PropTypes.bool,
  bgc: PropTypes.oneOf(["primary", "primary-light"])
};

export default Button;
