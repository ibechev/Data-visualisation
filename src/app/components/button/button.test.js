import React from "react";
import { mount } from "enzyme";
import Button from "./Button";
import renderer from "react-test-renderer";
import { propTypeErrors } from "../../../../tests/config/testUtils";

describe("Expect <Button />", () => {
  const childrenText = "Some text";
  const setup = (testProps = {}) => {
    const props = {
      ...testProps
    };
    return mount(
      <Button {...props}>
        <p>{childrenText}</p>
      </Button>
    );
  };

  it("to render correectly", () => {
    const wrapper = renderer
      .create(
        <Button>
          <p>Some text</p>
        </Button>
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("to not thrwo error with correct props", () => {
    propTypeErrors(
      <Button>
        <p>Some text</p>
      </Button>,
      {}
    );
    propTypeErrors(
      //eslint-disable-next-line
      <Button noBody onClick={() => {}}>
        <p>Some text</p>
      </Button>,
      { noBody: true, onClick: () => {} }
    );
    propTypeErrors(
      //eslint-disable-next-line
      <Button noBody onClick={() => {}}>
        <p>Some text</p>
      </Button>,
      { bgc: "primary", onClick: () => {} }
    );
    propTypeErrors(
      //eslint-disable-next-line
      <Button noBody onClick={() => {}}>
        <p>Some text</p>
      </Button>,
      { bgc: "primary-light", onClick: () => {} }
    );
  });

  describe("when prop 'noBody' is declared", () => {
    const wrapper = setup({ noBody: true });

    it("to set class 'no-body' when prop 'noBody' of the <button />", () => {
      expect(wrapper.find("button").hasClass("no-body")).toEqual(true);
    });

    it("to not set classes 'primary' or 'primary-light' of the <button />", () => {
      expect(wrapper.find("button").hasClass("primary")).toEqual(false);
      expect(wrapper.find("button").hasClass("primary-light")).toEqual(false);
    });
  });

  describe("when prop 'noBody' is not declared", () => {
    const wrapper = setup();

    it("to not set class 'no-body' of <button />", () => {
      expect(wrapper.find("button").hasClass("no-body")).toEqual(false);
    });

    it("to set class 'primary' when it's passed as prop 'bgc'", () => {
      expect(wrapper.find("button").hasClass("primary")).toEqual(true);
    });
  });

  it("to call prop 'onClick' when the button is clicked and prop 'onClick' has been passed as function", () => {
    const clickSpy = jest.fn();
    const wrapper = setup({ onClick: clickSpy });
    wrapper.find("button").simulate("click");
    expect(clickSpy).toHaveBeenCalledTimes(1);
  });

  it("to render any passed children", () => {
    const wrapper = setup();
    expect(wrapper.find("p").text()).toEqual(childrenText);
  });

  it("to set prop 'noBody' to 'false', 'bgc' to 'primary' and 'onClick' to 'null' by default (without any passed props)", () => {
    const wrapper = setup();
    expect(wrapper.props().bgc).toEqual("primary");
    expect(wrapper.props().noBody).toEqual(false);
    expect(wrapper.props().onClick).toEqual(null);
  });
});
