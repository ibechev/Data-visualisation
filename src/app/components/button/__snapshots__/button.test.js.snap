// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <Button /> to render correectly 1`] = `
<button
  className="button primary"
  onClick={[Function]}
  type="button"
>
  <p>
    Some text
  </p>
</button>
`;
